### ISSUE TITLE
Incorrect type error for method calls on union types

### ISSUE THREAD

**Issue by cyril0124** [EmmyLuaLs/emmylua-analyzer-rust#717]
The language server reports a false positive type error when a method is called on a variable that holds a union type. It seems to incorrectly infer the type of self.

<img width="588" height="220" alt="Image" src="https://github.com/user-attachments/assets/5cf27b2a-19a3-453d-b4fa-cf8eeac776b0" />

Note: This can also be recognized normally in LuaLS.

Can be avoided in the following way, but it's uncertain whether this error reporting behavior meets the design expectations of EmmyLuaLS.

<img width="428" height="221" alt="Image" src="https://github.com/user-attachments/assets/eed2c316-8381-4340-8997-0b87d2f3688e" />

**Comment by xuhuanzy** [comment]
luals是匹配任一就不管了, 但 emmylua 更严格, 不是很好解决

**Comment by xuhuanzy** [comment]
完整的可复现代码发一下

**Comment by cyril0124** [comment]
> 完整的可复现代码发一下

这个是存在 Bug 的代码：
```lua
---@class Class1
---@field eval fun(self: Class1)

---@class Class2
---@field eval fun(self: Class2)

---@class Class3
---@field v Class1|Class2
local c1 = {}


c1.v:eval()
```


这个是功能正常的代码：
```lua
---@class HasEval
---@field eval fun(self: HasEval)

---@class Class1: HasEval

---@class Class2: HasEval

---@class Class3
---@field v Class1|Class2
local c1 = {}

c1.v:eval()
```

**Comment by xuhuanzy** [comment]
已修复
